packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils_new.ts

```diff
@@ -0,0 +1,134 @@
+}
+
+export class BalancerV2SwapInfoCache extends SwapInfoCache {
+    // Options used when finding paths.
+    swapOptions: SwapOptions = {
+        gasPrice: parseFixed('50', 9),
+        swapGas: BigNumber.from('35000'),
+        poolTypeFilter: PoolFilter.All,
+        maxPools: 4,
+        timestamp: Math.floor(Date.now() / 1000),
+        forceRefresh: false,
+    };
+    routeProposer: RouteProposer;
+    poolDataService: SubgraphPoolDataService;
+
+    constructor(
+        chainId: ChainId,
+        private readonly subgraphUrl: string = BALANCER_V2_SUBGRAPH_URL_BY_CHAIN[chainId],
+        private readonly maxPoolsFetched: number = BALANCER_MAX_POOLS_FETCHED,
+        private readonly _topPoolsFetched: number = BALANCER_TOP_POOLS_FETCHED,
+        private readonly _warningLogger: LogFunction = DEFAULT_WARNING_LOGGER,
+        cache: { [key: string]: CacheValue } = {},
+    ) {
+        super(cache);
+        // SOR RouteProposer finds paths between a token pair using direct/multihop/linearPool routes (function will be exposed via SDK/SOR)
+        this.routeProposer = new RouteProposer();
+        // Uses Subgraph to retrieve up to date pool data (function will be exposed via SDK/SOR)
+        this.poolDataService = new SubgraphPoolDataService(this.subgraphUrl);
+        void this._loadTopPoolsAsync();
+        // Reload the top pools every 12 hours
+        setInterval(async () => void this._loadTopPoolsAsync(), ONE_DAY_MS / 2);
+    }
+
+    /**
+     * Retrieves swap steps and assets for a token pair.
+     * @param pool 
+     * @param takerToken 
+     * @param makerToken 
+     * @returns 
+    */
+    private _getPoolPairSwapInfo(pools: BalancerPoolResponse[], takerToken: string, makerToken: string): BalancerSwapInfo[] {
+        /*
+        Uses Balancer SOR/SDK functions to construct available paths for pair.
+        Paths can be direct, i.e. both tokens are in same pool or multihop.
+        Will also create paths for the new Balancer Linear pools.
+        */
+        // TO DO - This is just pseudo code to demonstrate potential option. Will need to expose functions from SOR/SDK.
+        // find available paths
+        const paths = this.routeProposer.getCandidatePaths(
+            takerToken,
+            makerToken,
+            SwapTypes.SwapExactIn,
+            pools,
+            this.swapOptions
+        );
+
+        // TO DO - Further filtering of these paths could be done using the already computed pathLimit which would determine if a path is valid for a specific trade amount.
+
+        if (paths.length == 0) return [];
+
+        // Helper function exposed from SDK/SOR that formats and returns swapSteps and assets
+        return formatSwaps(paths);
+    }
+
+    protected async _loadTopPoolsAsync(): Promise<void> {
+        const fromToSwapInfo: {
+            [from: string]: { [to: string]: BalancerSwapInfo[] };
+        } = {};
+
+        const pools = this.poolDataService.fetchPools();
```

Comment:
Does this make any blocking external requests?

Reply:
It would only be a subgraph query which I believe should be the same as the current `_fetchTopPoolsAsync()`.

Reply:
Apologies, I've missed out the `await` there. It will be `const pools = await this.poolDataService.fetchPools();`